wee_slack.py

```diff
@@ -1386,142 +1415,148 @@ def create_reaction_string(reactions):
+        if "subtype" in message_json and message_json["subtype"] in known_subtypes:
+            proc[message_json["subtype"]](message_json)
 
-        if "subtype" in message_json and message_json["subtype"] == "message_changed":
-                if "edited" in message_json["message"]:
-                    append = " (edited)"
-                else:
-                    append = ''
-                channel.change_message(message_json["message"]["ts"], text + append)
-                cache=False
-        elif "subtype" in message_json and message_json["subtype"] == "message_deleted":
-            append = "(deleted)"
-            text = ""
-            channel.change_message(message_json["deleted_ts"], text + append)
-            cache = False
-        elif message_json.get("subtype", "") == "channel_leave":
-            channel.buffer_prnt(w.prefix("quit").rstrip(), text, time)
-        elif message_json.get("subtype", "") == "channel_join":
-            channel.buffer_prnt(w.prefix("join").rstrip(), text, time)
-        elif message_json.get("subtype", "") == "channel_topic":
-            channel.buffer_prnt(w.prefix("network").rstrip(), text, time)
-        elif text.startswith("_") and text.endswith("_"):
-            text = text[1:-1]
-            if name != channel.server.nick:
-                text = name + " " + text
-            channel.buffer_prnt(w.prefix("action").rstrip(), text, time)
         else:
-            channel.buffer_prnt(name, text, time)
+            server = servers.find(message_json["_server"])
+            channel = channels.find(message_json["channel"])
+
+            #do not process messages in unexpected channels
+            if not channel.active:
+                channel.open(False)
+                dbg("message came for closed channel {}".format(channel.name))
+                return
+
+            time = message_json['ts']
+            text = render_message(message_json)
+            name = get_user(message_json, server)
+            name = name.encode('utf-8')
+
+            #special case with actions.
+            if text.startswith("_") and text.endswith("_"):
+                text = text[1:-1]
+                if name != channel.server.nick:
+                    text = name + " " + text
+                channel.buffer_prnt(w.prefix("action").rstrip(), text, time)
+
+            else:
+                channel.buffer_prnt(name, text, time)
 
-        if cache:
-            channel.cache_message(message_json)
+
+            if cache:
+                channel.cache_message(message_json)
 
     except Exception:
         if channel and ("text" in message_json) and message_json['text'] is not None:
             channel.buffer_prnt('unknown', message_json['text'])
         dbg("cannot process message {}\n{}".format(message_json, traceback.format_exc()))
 
-def unwrap_message(message_json):
+
+def process_message_changed(message_json):
```

Comment:
simplified process_message() - most important change is that we call render_message to do the manipulation of data and now cache the rendered version of the message so we don't need to call render_message again if we have already done it. Trades some memory for speed.